streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -875,4 +912,37 @@ private void initializeTransactions() {
             );
         }
     }
+
+    // visible for testing
+    String encodeTimestamp(final long partitionTime) {
+        final ByteBuffer buffer = ByteBuffer.allocate(9);
+        buffer.put(LATEST_MAGIC_BYTE);
+        buffer.putLong(partitionTime);
+        return Base64.getEncoder().encodeToString(buffer.array());
```

Comment:
@cadonna Yeah, it is still unclear at this point if the ```metadata``` field in ```OffsetAndMetadata``` could be used in this manner. @guozhangwang or @hachikuji knows this matter better. Anyhow, OffsetAndMetadata right now is the only medium through which we can checkpoint partition time anyways. So we might be stuck with using the ```metadata``` field.

Reply:
I don't have the full context on the history, but it would not be easy to change the API... I talked to Jason about it, and it seem we can just move forward with this PR as-is, and could do a KIP later that allows us to store metadata as `byte[]` type if we really need to change it. Atm, the metadata is just a few bytes and the overhead does not really matter IMHO.